src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -4348,6 +4279,85 @@ private static void EmitTimeoutCheckIfNeeded(IndentedTextWriter writer, RegexMet
             }
         }
 
+        /// <summary>Tries to create an IndexOf expression for the node.</summary>
+        /// <param name="node">The RegexNode. If it's a loop, only the one/notone/set aspect of the node is factored in.</param>
+        /// <param name="useLast">true to use LastIndexOf variants; false to use IndexOf variants.</param>
+        /// <param name="negate">true to search for the opposite of the node.</param>
+        /// <param name="literalLength">0 if returns false. If it returns true, string.Length for a multi, otherwise 1.</param>
+        /// <param name="indexOfExpr">The resulting expression if it returns true; otherwise, null.</param>
+        /// <returns>true if an expression could be produced; otherwise, false.</returns>
+        private static bool TryEmitIndexOf(
+            RegexNode node,
+            bool useLast, bool negate,
+            out int literalLength, [NotNullWhen(true)] out string? indexOfExpr)
+        {
+            string last = useLast ? "Last" : "";
+
+            if (node.Kind == RegexNodeKind.Multi)
+            {
+                Debug.Assert(!negate, "Negation isn't appropriate for a multi");
+                indexOfExpr = $"{last}IndexOf({Literal(node.Str)})";
+                literalLength = node.Str.Length;
+                return true;
+            }
+
+            if (node.IsOneFamily)
+            {
+                indexOfExpr = negate ? $"{last}IndexOfAnyExcept({Literal(node.Ch)})" : $"{last}IndexOf({Literal(node.Ch)})";
+                literalLength = 1;
+                return true;
+            }
+
+            if (node.IsNotoneFamily)
+            {
+                indexOfExpr = negate ? $"{last}IndexOf({Literal(node.Ch)})" : $"{last}IndexOfAnyExcept({Literal(node.Ch)})";
+                literalLength = 1;
+                return true;
+            }
+
+            if (node.IsSetFamily)
+            {
+                bool negated = RegexCharClass.IsNegated(node.Str) ^ negate;
+
+                Span<char> setChars = stackalloc char[5]; // current max that's vectorized
```

Comment:
> I know it precedes this PR, but should we be guarding against 5 if we're not using IndexOfAny? The rest of the *Any* helpers are vectorized up to 4.

It's not 100% correct, but it's also a reasonable limit that prevents the code complexity from getting out of hand.

> Once https://github.com/dotnet/runtime/issues/68328 is available, should we defer such cases to it, as it can make the optimal runtime decision?

I'm very hesitant to do that.  The source generator can generate very efficient loops and set tests at compile time.  I expect it would end up being a deoptimization, both in startup time and in throughput, to always delegate to IndexOfAny.

Reply:
> to always delegate to IndexOfAny

By always, do you mean for all sets of 4 & 5 characters?

Reply:
I mean for anything we're not sure it's going to optimize / vectorize.  What if those 4 or 5 characters aren't ASCII?

Reply:
> What if those 4 or 5 characters aren't ASCII? Will the probabilistic map be employed / precomputed?

Depends on how many specialized cases we wanted to implement, but judging by my current local implementation it would mean using the 4 value vectorized helper when not on Mono, otherwise deferring to a simple loop / the probabilistic map.

> Will it necessarily be better? And if so, would it not be better for the source generator to precompute that?

Currently cases like `span.IndexOfAnyExcept(4 values)` will go to [`ProbabilisticMap.IndexOfAny`](https://github.com/dotnet/runtime/blob/fb001c47d8fe6575c390fa7fc47b039917882841/src/libraries/System.Private.CoreLib/src/System/ProbabilisticMap.cs#L103), which may use a for loop or spend a few extra cycles for the init cost of the ASCII bitmap/probabilistic map on every call.
`IndexOfAnyValues` would be faster in this case as we pay the init cost upfront.

The source generator could of course do as good/better of a job at the cost of increasing the amount of code it emits.

Reply:
> Depends on how many specialized cases we wanted to implement

If we special-case 4 and 5 inputs with a concrete type that can be devirtualized and just delegate to the internal 4/5-overload versions, I'm fine removing the reliance on the 4/5-based IndexOfAny overload and just using the new one for those as well.

> when not on Mono

The Mono situation is temporary.  It resulted from a last minute revert due to lack of generic specialization in certain circumstances.  The same pattern is used by the new IndexOfAny vectorized changes.  I expect the issue to be fixed in Mono (it may already be) and the special-casing for Mono reverted.